src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -35,12 +35,12 @@ class TaskReplaceActor(
 
   // compute all values ====================================================================================
 
-  // All running instances of this app
+  // All existing instances of this app independent of the goal.
   //
   // Killed resident tasks are not expunged from the instances list. Ignore
   // them. LaunchQueue takes care of launching instances against reservations
   // first
-  val currentInstances = instanceTracker.specInstancesSync(runSpec.id).filter(_.state.goal == Goal.Running)
+  val currentInstances = instanceTracker.specInstancesSync(runSpec.id)
```

Comment:
Bug nr. 1: we would only take `Running` instances into consideration, however, during an upgrade there would be old instances in the process of being killed with a `Goal.Decommissioned`

Reply:
But these should not be kill, right? As far as I know the `currentInstances` also select the old instances that should receive a kill.

Reply:
Come to think of it I think we should update this instance set on each update and then call killing, launching and finished check. That would simple the logic quite a bit. It would break some unit tests and require a little refactoring in the starting behaviour.

Reply:
Yeah, should be a separate effort